lambda_function.py

```diff
@@ -0,0 +1,48 @@
+import os
+import json
+import urllib3
+import boto3
+from urllib3.exceptions import ConnectTimeoutError
+from boto3 import resource
+from botocore.exceptions import ClientError
+from boto3.dynamodb.conditions import Key, Attr
+
+http = urllib3.PoolManager()
+client = boto3.client('events', region_name="eu-west-2")
+
+
+def get_item_with_id(table, primary_key, primary_key_value):
+    response = table.query(KeyConditionExpression=Key(primary_key).eq(primary_key_value))
+    items = response["Items"]
+
+    if len(items) == 0:
```

Comment:
Don't throw an error in this case, just no-op. If a user makes a metadata change within the Preservation System the entity will be updated in CC and a message sent, but there will be no items in our ingest table.

Reply:
Ah, ok. Since the notification includes whether it's a `Created` or `Updated` item and whether it's `Metadata` or `File`, I'll instead only throw an error if it's `Created` And `File`;  that way, we'll know if something has truly gone wrong

Reply:
@kurtismash In addition to this, I've thought of different scenarios and the actions to take, let me know if this is correct:

1. Metadata Creation message, file does not exist in table - **no op**
2. File Creation message, file does not exist in table - **no op**
3. Metadata Creation message, file exists in table - **no op**
4. File Creation message, file exists in table  - **`ingested_CC` = true**
5. Metadata Update message - **no op**
6. File Update message - **no op**

Reply:
I don't think I know enough about the messages being produced by the disaster recovery worker to answer these. In its simplest form, every time the DR worker processes an entity it needs to send a message to say it's processed that entity, and that message needs to include the SourceID/bitstream uuid from Preservica. This Lambda then acts on these messages, querying the table using the SourceID/bitstream uuid and writing `ingested_CC = true` for any items that are returned.

Reply:
Yes, but once `ingested_CC = true` exists, subsequent messages with this id in the future, can just be ignored so long as `ingested_CC = true` exists, so maybe before calling `update_item()` I could add a check to see if it's already true and then do nothing if it is?

Reply:
We should test if making an update call that doesn't result in a change causes the item to re-appear in the DDB stream. I'm not too worried if things happen more than once as long as we're eventually consistent, this includes us sending notification messages to other teams. Until someone complains the only benefit would be a cost saving.

However, you do already have the field in memory after your query so it won't cost anything more to retrieve it. Now you mention it, there's a good chance in future that items will stay in this table longer which would cause us to send multiple of the same notification as DR processes. Let's do the check.